crates/bevy_platform_support/src/staging.rs

```diff
@@ -0,0 +1,460 @@
+pub struct StagerLocked<'a, T: StagedChanges, C: ColdStorage<T>> {
+    cold: C,
+    staged: RwLockWriteGuard<'a, T>,
+}
+
+/// A locked version of [`StagedRef`] Use this to hold a lock guard while using [`StagerLocked::as_staged_ref`].
+pub struct StagedRefLocked<'a, T: StagedChanges, C: ColdStorage<T>> {
+    cold: C,
+    staged: RwLockReadGuard<'a, T>,
+}
+
+/// A general purpose enum for representing data that may or may not need to be staged.
+pub enum MaybeStaged<C, S> {
+    /// There is no staging necessary.
+    Cold(C),
+    /// There is staging necessary.
+    Staged(S),
+}
+
+/// A struct that allows read-optimized operations while still allowing mutation. When mutattions are made,
+/// they are staged. Then, at user-defined times, they are applied to the read-optimized storage. This allows muttations
+/// through [`RwLock`]s without needing to constantly lock old or cold data.
+///
+/// This is not designed for atomic use (ie. in an [`Arc`]). See [`AtomicStageOnWrite`] for that.
+#[derive(Default)]
+pub struct StageOnWrite<T: StagedChanges> {
+    /// Cold data is read optimized.
+    cold: T::Cold,
+    /// Staged data stores recent modifications to cold. It's [`RwLock`] coordinates mutations.
+    staged: RwLock<T>,
+}
+
+#[derive(Default)]
+struct AtomicStageOnWriteInner<T: StagedChanges> {
+    /// Cold data is read optimized. This lives behind a [`RwLock`], but it is only written to for applying changes in
+    /// a non-blocking way. In other worlds this locks, but almost never blocks. (It can technically block if a thread
+    /// tries to read from it while it is having changes applied, but that is extremely rare.)
+    cold: RwLock<T::Cold>,
+    /// Staged data stores recent modifications to cold. It's [`RwLock`] coordinates mutations.
+    staged: RwLock<T>,
+}
+
+/// A version of [`StageOnWrite`] designed for atomic use. See [`StageOnWrite`] for details.
+///
+/// This type includes a baked in [`Arc`], so it can be shared similarly. Many of it's methods take `&mut self` even though
+/// it doesn't technically need the mutation. This is done to signify that the methods involve a state change of the data and to prevent deadlocks.
+/// Because everything that involves a write lock requires `&mut self`, it is impossible to deadlock because doing so would require another lock guard
+/// with the same lifetime, which rust will complaine about. If you do not want this behavior, see [`AtomicStageOnWriteInner`].
+#[derive(Clone)]
+pub struct AtomicStageOnWrite<T: StagedChanges>(Arc<AtomicStageOnWriteInner<T>>);
+
+impl<T: StagedChanges> StageOnWrite<T> {
+    /// Constructs a new [`StageOnWrite`] with the given value and no staged changes.
+    pub fn new(value: T::Cold) -> Self {
+        Self {
+            cold: value,
+            staged: RwLock::default(),
+        }
+    }
+
+    /// Gets the inner cold data if it is safe. If [`any_staged`](Self::any_staged) is known to be false, this can be safely unwrapped.
```

Comment:
```suggestion
    /// Gets the inner cold data if it is safe.
    /// If [`any_staged`](Self::any_staged) is known to be false, this can be safely unwrapped.
```
What does _"if it is safe"_ refer to here? I generally avoid using the term "safe" in Rust documentation unless it is explicitly around `unsafe`/safety.